File: 2025_Oct/Anica/week4/2025-11-10-car-fleet.py
class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        pair = [[p, s] for p, s in zip(position, speed)]
        pair.sort()
        stack = []

        for p, s in pair[::-1]:
            stack.append((target-p)/s)
            if len(stack) >= 2: 
                if stack[-1] <= stack[-2]:
                    stack.pop() 
        return len(stack)

File: 2025_Oct/evelyn/week4/2025-11-10-car-fleet.cpp
class Solution {
public:
    int carFleet(int target, vector<int>& position, vector<int>& speed) {

        vector<pair<int, double>> cars;

        for (int i = 0; i < position.size(); i++) {
            double time = (double)(target - position[i]) / speed[i];
            cars.push_back({position[i], time});
        }

        sort(cars.begin(), cars.end(), [](auto &a, auto &b){
            return a.first > b.first;
        });

        int fleets = 0;
        double slowTime = 0;

        for (auto &c : cars) {
            if (c.second > slowTime) {
                fleets++;
                slowTime = c.second;
            }
        }

        return fleets;
    }
};
